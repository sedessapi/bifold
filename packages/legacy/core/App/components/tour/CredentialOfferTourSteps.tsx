import * as React from 'react'
import { useTranslation } from 'react-i18next'
import { Text } from 'react-native'

import { useTheme } from '../../contexts/theme'
import { TourStep } from '../../contexts/tour/tour-context'

import { TourBox } from './TourBox'

export const credentialOfferTourSteps: TourStep[] = [
  {
    render: (props) => {
      const { currentTour, currentStep, next, stop, previous } = props
      const { t } = useTranslation()
      const { ColorPallet, TextTheme } = useTheme()
      return (
        <TourBox
          title={t('Tour.CredentialOffers')}
          hideLeft
          rightText={t('Tour.Done')}
          onRight={stop}
          currentTour={currentTour}
          currentStep={currentStep}
          previous={previous}
          stop={stop}
          next={next}
        >
          <Text
            style={{
              ...TextTheme.normal,
              color: ColorPallet.notification.infoText,
            }}
            allowFontScaling={false}
          >
            {t('Tour.CredentialOffersDescription')}
          </Text>
        </TourBox>
      )
    },
  },
]
