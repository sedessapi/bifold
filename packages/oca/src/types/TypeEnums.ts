export enum CaptureBaseAttributeType {
  Binary = 'Binary',
  Text = 'Text',
  DateTime = 'DateTime',
  Numeric = 'Numeric',
  DateInt = 'DateInt',
}

export enum OverlayType {
  CaptureBase10 = 'spec/capture_base/1.0',
  Meta10 = 'spec/overlays/meta/1.0',
  Label10 = 'spec/overlays/label/1.0',
  Format10 = 'spec/overlays/format/1.0',
  CharacterEncoding10 = 'spec/overlays/character_encoding/1.0',
  Branding01 = 'aries/overlays/branding/0.1',
  Branding10 = 'aries/overlays/branding/1.0',
}
