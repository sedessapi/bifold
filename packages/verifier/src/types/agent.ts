import {
  AnonCredsCredentialFormatService,
  AnonCredsModule,
  AnonCredsProofFormatService,
  LegacyIndyCredentialFormatService,
  LegacyIndyProofFormatService,
  V1CredentialProtocol,
  V1ProofProtocol,
} from '@aries-framework/anoncreds'
import {
  Agent,
  ConnectionsModule,
  CredentialsModule,
  MediationRecipientModule,
  ProofsModule,
  V2CredentialProtocol,
  V2ProofProtocol,
} from '@aries-framework/core'

function getAgentModules() {
  return null as unknown as {
    //askar: AskarModule
    //anoncredsRs: AnonCredsRsModule
    anoncreds: AnonCredsModule
    //indyVdr: IndyVdrModule
    connections: ConnectionsModule
    credentials: CredentialsModule<
      (
        | V1CredentialProtocol
        | V2CredentialProtocol<(LegacyIndyCredentialFormatService | AnonCredsCredentialFormatService)[]>
      )[]
    >
    proofs: ProofsModule<
      (V1ProofProtocol | V2ProofProtocol<(LegacyIndyProofFormatService | AnonCredsProofFormatService)[]>)[]
    >
    mediationRecipient: MediationRecipientModule
  }
}

export type BifoldAgent = Agent<ReturnType<typeof getAgentModules>>
