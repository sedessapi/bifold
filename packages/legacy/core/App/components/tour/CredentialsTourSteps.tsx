import * as React from 'react'
import { useTranslation } from 'react-i18next'
import { Text } from 'react-native'

import { useTheme } from '../../contexts/theme'
import { TourStep } from '../../contexts/tour/tour-context'

import { TourBox } from './TourBox'

export const credentialsTourSteps: TourStep[] = [
  {
    render: (props) => {
      const { currentTour, currentStep, next, stop, previous } = props
      const { t } = useTranslation()
      const { ColorPallet, TextTheme } = useTheme()
      return (
        <TourBox
          title={t('Tour.AddCredentials')}
          hideLeft
          rightText={t('Tour.Done')}
          onRight={stop}
          currentTour={currentTour}
          currentStep={currentStep}
          previous={previous}
          stop={stop}
          next={next}
        >
          <Text
            style={{
              ...TextTheme.normal,
              color: ColorPallet.notification.infoText,
            }}
            allowFontScaling={false}
          >
            {t('Tour.AddCredentialsDescription')}
          </Text>
        </TourBox>
      )
    },
  },
]
