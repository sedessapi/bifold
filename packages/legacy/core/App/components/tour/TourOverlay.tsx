import React, { useCallback, useContext, useRef, useState, useEffect } from 'react'
import { ColorValue, LayoutRectangle, Modal, View, ViewStyle, useWindowDimensions } from 'react-native'
import { Defs, Mask, Rect, Svg } from 'react-native-svg'

import { tourMargin } from '../../constants'
import { BackdropPressBehavior, OSConfig, TourContext, TourStep } from '../../contexts/tour/tour-context'
import { Optional, TourID } from '../../types/tour'
import { testIdWithKey } from '../../utils/testable'

import { SpotCutout } from './SpotCutout'

interface TourOverlayProps {
  backdropOpacity: number
  color: ColorValue
  currentTour: TourID
  currentStep: Optional<number>
  nativeDriver: boolean | OSConfig<boolean>
  onBackdropPress: Optional<BackdropPressBehavior>
  changeSpot: (spot: LayoutRectangle) => void
  spot: LayoutRectangle
  tourStep: TourStep
}

export const TourOverlay = (props: TourOverlayProps) => {
  const { width: windowWidth, height: windowHeight } = useWindowDimensions()
  const { color, currentTour, currentStep, onBackdropPress, backdropOpacity, changeSpot, spot, tourStep } = props

  const { next, previous, start, stop } = useContext(TourContext)

  const [tooltipStyle, setTooltipStyle] = useState<ViewStyle>({})

  const tooltipRef = useRef<View>(null)

  const handleBackdropPress = useCallback((): void => {
    const handler = tourStep.onBackdropPress ?? onBackdropPress

    if (handler !== undefined && currentStep !== undefined) {
      switch (handler) {
        case 'continue':
          return next()

        case 'stop':
          return stop()

        default:
          return handler({ currentTour, currentStep, changeSpot, next, previous, start, stop })
      }
    }
  }, [tourStep, onBackdropPress, currentTour, currentStep, changeSpot, next, previous, start, stop])

  useEffect(() => {
    const gapBetweenSpotAndTooltip = 50
    // if origin spot (ie. no spotlight)
    if (spot.x === 0 && spot.y === 0) {
      const top = windowHeight / 5
      setTooltipStyle({
        left: tourMargin,
        right: tourMargin,
        top,
      })
      // if spot is in the lower half of the screen
    } else if (spot.y >= windowHeight / 2) {
      const bottom = windowHeight - spot.y + gapBetweenSpotAndTooltip
      setTooltipStyle({
        left: tourMargin,
        right: tourMargin,
        bottom,
      })
      // if spot is in the upper half of the screen
    } else {
      const top = spot.y + gapBetweenSpotAndTooltip + spot.height
      setTooltipStyle({
        left: tourMargin,
        right: tourMargin,
        top,
      })
    }
  }, [spot.height, spot.width, spot.x, spot.y])

  return (
    <Modal
      animationType="fade"
      presentationStyle="overFullScreen"
      transparent={true}
      visible={currentStep !== undefined}
    >
      <View style={{ height: windowHeight, width: windowWidth }} testID={testIdWithKey('SpotlightOverlay')}>
        <Svg
          testID={testIdWithKey('SpotOverlay')}
          height="100%"
          width="100%"
          viewBox={`0 0 ${windowWidth} ${windowHeight}`}
          onPress={handleBackdropPress}
          shouldRasterizeIOS={true}
          renderToHardwareTextureAndroid={true}
        >
          <Defs>
            <Mask id="mask" x={0} y={0} height="100%" width="100%">
              <Rect height="100%" width="100%" fill="#fff" />
              <SpotCutout />
            </Mask>
          </Defs>

          <Rect height="100%" width="100%" fill={color} mask="url(#mask)" opacity={backdropOpacity} />
        </Svg>

        <View
          ref={tooltipRef}
          testID={testIdWithKey('SpotTooltip')}
          style={{ ...tooltipStyle, opacity: 1, position: 'absolute' }}
        >
          {currentStep !== undefined && (
            <tourStep.render
              currentTour={currentTour}
              currentStep={currentStep}
              changeSpot={changeSpot}
              next={next}
              previous={previous}
              stop={stop}
            />
          )}
        </View>
      </View>
    </Modal>
  )
}
