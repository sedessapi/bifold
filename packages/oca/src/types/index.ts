import OverlayTypeMap from './OverlayTypeMap'
import { OverlayType, CaptureBaseAttributeType } from './TypeEnums'
import BaseOverlay from './base/BaseOverlay'
import BrandingOverlay from './branding/BrandingOverlay'
import LegacyBrandingOverlay from './branding/LegacyBrandingOverlay'
import OverlayBundle from './bundle/OverlayBundle'
import CaptureBase from './capture-base/CaptureBase'
import CharacterEncodingOverlay from './semantic/CharacterEncodingOverlay'
import FormatOverlay from './semantic/FormatOverlay'
import InformationOverlay from './semantic/InformationOverlay'
import LabelOverlay from './semantic/LabelOverlay'
import MetaOverlay from './semantic/MetaOverlay'
import StandardOverlay from './semantic/StandardOverlay'

export {
  OverlayType,
  CaptureBaseAttributeType,
  BaseOverlay,
  CaptureBase,
  OverlayBundle,
  BrandingOverlay,
  LegacyBrandingOverlay,
  CharacterEncodingOverlay,
  FormatOverlay,
  InformationOverlay,
  LabelOverlay,
  MetaOverlay,
  StandardOverlay,
  OverlayTypeMap,
}
