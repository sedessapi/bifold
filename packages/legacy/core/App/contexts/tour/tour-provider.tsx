import React, { forwardRef, useCallback, useImperativeHandle, useMemo, useState } from 'react'
import { ColorValue, LayoutRectangle } from 'react-native'

import { TourOverlay } from '../../components/tour/TourOverlay'
import { ChildFn, TourID } from '../../types/tour'
import { isChildFunction } from '../../utils/helpers'

import { BackdropPressBehavior, OSConfig, Tour, TourContext, TourCtx, TourStep, ORIGIN_SPOT } from './tour-context'

export interface TourProviderProps {
  children: React.ReactNode | ChildFn<Tour>
  /**
   * Define if the animations in the tour should use the native driver or not.
   * A boolean can be used to apply the same value to both Android and iOS, or
   * an object with `android` and `ios` keys can be used to define a value for
   * each OS.
   *
   * @default false
   */
  nativeDriver?: boolean | OSConfig<boolean>
  /**
   * Sets the default behavior of pressing the tour's backdrop. You can use
   * either one of the following values:
   * - A callback function with the {@link Tour} options object in the
   * first argument. This allows more franular control over the tour.
   * - The `continue` literal string, which is a shortcut to move to the next
   * step, and stop the tour on the last step.
   * - the `stop` literal string, which is a shortcut to stop the tour.
   *
   * **NOTE:** You can also override this behavior on each step configuration.
   */
  onBackdropPress?: BackdropPressBehavior
  /**
   * The color of the overlay of the tour.
   *
   * @default black
   */
  overlayColor?: ColorValue
  /**
   * The opacity applied to the overlay of the tour (between 0 to 1).
   *
   * @default 0.45
   */
  overlayOpacity?: number
  /**
   * The list of steps for the home tour.
   */
  homeTourSteps: TourStep[]
  /**
   * Same as above for the credential list screen
   */
  credentialsTourSteps: TourStep[]
  /**
   * Same as above for the credential offer screen
   */
  credentialOfferTourSteps: TourStep[]
  /**
   * Same as above for the proof request screen
   */
  proofRequestTourSteps: TourStep[]
}

export const TourProvider = forwardRef<Tour, TourProviderProps>((props, ref) => {
  const {
    children,
    onBackdropPress,
    overlayColor = 'black',
    overlayOpacity = 0.45,
    homeTourSteps,
    credentialsTourSteps,
    credentialOfferTourSteps,
    proofRequestTourSteps,
    nativeDriver = false,
  } = props

  const [currentTour, setCurrentTour] = useState<TourID>(TourID.HomeTour)
  const [currentStep, setCurrentStep] = useState<number>()
  const [spot, setSpot] = useState(ORIGIN_SPOT)

  const renderStep = useCallback(
    (index: number): void | Promise<void> => {
      if (
        (currentTour === TourID.HomeTour && homeTourSteps[index] !== undefined) ||
        (currentTour === TourID.CredentialsTour && credentialsTourSteps[index] !== undefined) ||
        (currentTour === TourID.CredentialOfferTour && credentialOfferTourSteps[index] !== undefined) ||
        (currentTour === TourID.ProofRequestTour && proofRequestTourSteps[index] !== undefined)
      ) {
        setCurrentStep(index)
      }
    },
    [currentTour, homeTourSteps, credentialsTourSteps, credentialOfferTourSteps, proofRequestTourSteps]
  )

  const changeSpot = useCallback((newSpot: LayoutRectangle): void => {
    setSpot(newSpot)
  }, [])

  const start = useCallback(
    (tourId: TourID): void => {
      setCurrentTour(tourId)
      renderStep(0)
    },
    [renderStep]
  )

  const stop = useCallback((): void => {
    setCurrentStep(undefined)
    setSpot(ORIGIN_SPOT)
  }, [])

  const next = useCallback((): void => {
    let steps = homeTourSteps
    if (currentTour === TourID.CredentialsTour) {
      steps = credentialsTourSteps
    } else if (currentTour === TourID.CredentialOfferTour) {
      steps = credentialOfferTourSteps
    } else if (currentTour === TourID.ProofRequestTour) {
      steps = proofRequestTourSteps
    }

    if (currentStep !== undefined) {
      currentStep === steps.length - 1 ? stop() : renderStep(currentStep + 1)
    }
  }, [
    stop,
    renderStep,
    currentStep,
    currentTour,
    homeTourSteps,
    credentialsTourSteps,
    credentialOfferTourSteps,
    proofRequestTourSteps,
  ])

  // works the same regardless of which tour is on
  const previous = useCallback((): void => {
    if (currentStep !== undefined && currentStep > 0) {
      renderStep(currentStep - 1)
    }
  }, [renderStep, currentStep])

  const tourStep = useMemo((): TourStep => {
    let stepToRender = undefined
    let steps = homeTourSteps
    if (currentTour === TourID.CredentialsTour) {
      steps = credentialsTourSteps
    } else if (currentTour === TourID.CredentialOfferTour) {
      steps = credentialOfferTourSteps
    } else if (currentTour === TourID.ProofRequestTour) {
      steps = proofRequestTourSteps
    }

    stepToRender = currentStep !== undefined ? steps[currentStep] : undefined
    return stepToRender ?? { render: () => <></> }
  }, [homeTourSteps, credentialsTourSteps, credentialOfferTourSteps, proofRequestTourSteps, currentStep])

  const tour = useMemo(
    (): TourCtx => ({
      changeSpot,
      currentTour,
      currentStep,
      next,
      previous,
      spot,
      start,
      stop,
      homeTourSteps,
      credentialsTourSteps,
      credentialOfferTourSteps,
      proofRequestTourSteps,
    }),
    [
      changeSpot,
      currentTour,
      currentStep,
      next,
      previous,
      spot,
      start,
      stop,
      homeTourSteps,
      credentialsTourSteps,
      credentialOfferTourSteps,
      proofRequestTourSteps,
    ]
  )

  useImperativeHandle(ref, () => ({
    currentTour,
    currentStep,
    changeSpot,
    next,
    previous,
    start,
    stop,
  }))

  return (
    <TourContext.Provider value={tour}>
      {isChildFunction(children) ? <TourContext.Consumer>{children}</TourContext.Consumer> : <>{children}</>}

      <TourOverlay
        color={overlayColor}
        currentStep={currentStep}
        currentTour={currentTour}
        changeSpot={changeSpot}
        backdropOpacity={overlayOpacity}
        onBackdropPress={onBackdropPress}
        spot={spot}
        tourStep={tourStep}
        nativeDriver={nativeDriver}
      />
    </TourContext.Provider>
  )
})
