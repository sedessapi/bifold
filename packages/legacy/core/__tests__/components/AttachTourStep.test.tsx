import { render } from '@testing-library/react-native'
import React from 'react'
import { View } from 'react-native'

import { AttachTourStep } from '../../App/components/tour/AttachTourStep'
import { credentialOfferTourSteps } from '../../App/components/tour/CredentialOfferTourSteps'
import { credentialsTourSteps } from '../../App/components/tour/CredentialsTourSteps'
import { homeTourSteps } from '../../App/components/tour/HomeTourSteps'
import { proofRequestTourSteps } from '../../App/components/tour/ProofRequestTourSteps'
import { TourProvider } from '../../App/contexts/tour/tour-provider'
import { TourID } from '../../App/types/tour'

describe('AttachTourStep', () => {
  test('Renders properly with defaults', () => {
    const tree = render(
      <TourProvider
        homeTourSteps={homeTourSteps}
        credentialsTourSteps={credentialsTourSteps}
        credentialOfferTourSteps={credentialOfferTourSteps}
        proofRequestTourSteps={proofRequestTourSteps}
        overlayColor={'gray'}
        overlayOpacity={0.7}
      >
        <AttachTourStep index={0} tourID={TourID.HomeTour}>
          <View />
        </AttachTourStep>
      </TourProvider>
    )

    expect(tree).toMatchSnapshot()
  })
})
