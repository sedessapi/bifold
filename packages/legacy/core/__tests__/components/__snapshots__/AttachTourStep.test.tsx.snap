// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachTourStep Renders properly with defaults 1`] = `
Array [
  <View
    collapsable={false}
    focusable={false}
    style={
      Object {
        "alignSelf": "center",
      }
    }
    testID="com.ariesbifold:id/AttachTourStep"
  >
    <View />
  </View>,
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    presentationStyle="overFullScreen"
    transparent={true}
    visible={false}
  />,
]
`;
